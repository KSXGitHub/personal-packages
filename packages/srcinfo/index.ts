import { Result, Ok, Err, ok, err } from '@tsfun/result'
export type { Result, Ok, Err }

const requiredFields = [
  'pkgver',
  'pkgrel',
  'epoch',
] as const

const partialFields = [
  'pkgdesc',
  'url',
  'install',
  'changelog',
] as const

const arrayFields = [
  'arch',
  'groups',
  'license',
  'noextract',
  'options',
  'backup',
  'validgpgkeys',
  'source',
  'depends',
  'checkdepends',
  'makedepends',
  'optdepends',
  'provides',
  'conflicts',
  'replaces',
  'md5sums',
  'sha1sums',
  'sha224sums',
  'sha256sums',
  'sha348sums',
  'sha512sums',
] as const

/** Field that always appear in `.SRCINFO` */
export type RequiredField = typeof requiredFields[number]
/** Field that appears at most once in `.SRCINFO` */
export type PartialField = typeof partialFields[number]
/** Field that may appear multiple times in `.SRCINFO` */
export type ArrayField = typeof arrayFields[number]

/** Interface of `.SRCINFO` */
export type SrcInfo = Readonly<
  & Record<RequiredField, string>
  & Partial<Record<PartialField, string>>
  & Record<ArrayField, readonly string[]>
>

function parseLine(line: string): readonly [string, string] | null {
  const result = /(?<key>[^\s]+)\s*=\s*(?<value>[^\s]+)/.exec(line)
  if (!result) return null
  const { key, value } = result.groups!
  return [key, value]
}

type SrcInfoLines = ReadonlyArray<readonly [string, string]>

const getUnique = (
  source: SrcInfoLines,
  key: RequiredField | PartialField,
) => source.find(pair => key === pair[0])?.[1] ?? null

const getArray = (
  source: SrcInfoLines,
  key: ArrayField,
) => source.filter(pair => key === pair[0]).map(pair => pair[1])

export type SrcInfoError =
  & Readonly<
    | { type: 'Syntax', line: string, index: number }
    | { type: 'MissingRequiredField', field: RequiredField }
  >
  & { readonly message: string }

/** Parse content of a `.SRCINFO` file */
export function parseSrcInfo(source: string): Result<SrcInfo, readonly SrcInfoError[]> {
  const lines = source
    .split('\n') // LF is sufficient, no needs for CRLF
    .map((line, index) => ({ line, index }))
    .filter(x => x.line.trim())
    .map(({ line, index }) => ({ line, index, entry: parseLine(line) }))

  const errors: SrcInfoError[] = []

  for (const { line, index, entry } of lines) {
    if (!entry) {
      errors.push({
        type: 'Syntax',
        line,
        index,
        message: `Line ${index + 1} is invalid`,
      })
    }
  }

  for (const field of requiredFields) {
    if (!lines.some(x => field === x.entry![0])) {
      errors.push({
        type: 'MissingRequiredField',
        field,
        message: `Field "${field}" is required but not found`,
      })
    }
  }

  if (errors.length) return err(errors)

  const parsedSource: SrcInfoLines = lines.map(x => x.entry!)

  const uniques = Object.fromEntries(
    [...requiredFields, ...partialFields].map(field => [field, getUnique(parsedSource, field)]),
  )

  const arrays = Object.fromEntries(
    arrayFields.map(field => [field, getArray(parsedSource, field)]),
  )

  return ok({ ...uniques, ...arrays } as any)
}
