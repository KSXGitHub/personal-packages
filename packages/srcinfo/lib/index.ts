export * from './parse-src-info'
export * from './types'
