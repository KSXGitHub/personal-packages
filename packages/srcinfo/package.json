{
  "name": "@khai96x/srcinfo",
  "version": "0.0.5",
  "description": "Parsing ArchLinux's .SRCINFO",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://github.com/KSXGitHub/personal-packages#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/KSXGitHub/personal-packages.git"
  },
  "bugs": {
    "url": "https://github.com/KSXGitHub/personal-packages/issues"
  },
  "keywords": [
    "pkgbuild",
    "archlinux",
    "aur",
    "srcinfo"
  ],
  "main": "index.js",
  "module": "index.mjs",
  "browser": "index.mjs",
  "bin": {
    "srcinfo-to-json": "./bin/srcinfo-to-json"
  },
  "sideEffects": false,
  "dependencies": {
    "yargs": "^17.0.0",
    "tslib": "^2.0.1",
    "@tsfun/result": "^0.0.20",
    "@types/yargs": "^15.0.5",
    "@types/node": "^14.6.4"
  }
}
