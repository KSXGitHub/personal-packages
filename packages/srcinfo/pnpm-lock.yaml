dependencies:
  '@tsfun/result': 0.0.20
  '@types/node': 14.6.4
  '@types/yargs': 15.0.5
  tslib: 2.0.1
  yargs: 16.0.0
lockfileVersion: 5.1
packages:
  /@tsfun/prv-option-result-common/0.0.15:
    dependencies:
      '@types/node': 14.6.0
      tslib: 2.0.1
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-oqmmMHkoD1IXADxtOupJ9JFBcocuToxbSuqMIw3VszZIyvysJ3t9y+LlP+qxo92pOj/GJz4oohhzKBHkaaLqWA==
  /@tsfun/prv-types/0.0.10:
    dependencies:
      '@types/node': 14.6.0
      tslib: 2.0.1
    dev: false
    resolution:
      integrity: sha512-8QJUR81Zyf9gLhl7llp10bTZ+4WgXBM6LaiKZTmgjmviggFIQobm1hDzirZ9gUMxG1lGtTr4glszSkvuCcSjaA==
  /@tsfun/result/0.0.20:
    dependencies:
      '@tsfun/prv-option-result-common': 0.0.15
      '@tsfun/prv-types': 0.0.10
      '@tsfun/tap': 0.0.14
      '@types/node': 14.6.0
      tslib: 2.0.1
    dev: false
    engines:
      node: '>= 8.9.0'
    resolution:
      integrity: sha512-7IXm8ab6hRl7dNliHvFCoHdDXcDITSOPFKZl+7VxEKTXXmSrUe5QupvzzBro1G5Q104do/xiHZT1RBhwLzg1iw==
  /@tsfun/tap/0.0.14:
    dependencies:
      '@types/node': 14.6.0
      tslib: 2.0.1
    dev: false
    resolution:
      integrity: sha512-hXrhnrFidsbi4zeo6HaiHEMFLcjNTSc4osYps2Py7cgbgwW6nNxcZwRbDlQaMIdvpuQuHFd1+bxhDYQ0C9r6xA==
  /@types/color-name/1.1.1:
    dev: false
    resolution:
      integrity: sha512-rr+OQyAjxze7GgWrSaJwydHStIhHq2lvY3BOC2Mj7KnzI7XK0Uw1TOOdI9lDoajEbSWLiYgoo4f1R51erQfhPQ==
  /@types/node/14.6.0:
    dev: false
    resolution:
      integrity: sha512-mikldZQitV94akrc4sCcSjtJfsTKt4p+e/s0AGscVA6XArQ9kFclP+ZiYUMnq987rc6QlYxXv/EivqlfSLxpKA==
  /@types/node/14.6.4:
    dev: false
    resolution:
      integrity: sha512-Wk7nG1JSaMfMpoMJDKUsWYugliB2Vy55pdjLpmLixeyMi7HizW2I/9QoxsPCkXl3dO+ZOVqPumKaDUv5zJu2uQ==
  /@types/yargs-parser/15.0.0:
    dev: false
    resolution:
      integrity: sha512-FA/BWv8t8ZWJ+gEOnLLd8ygxH/2UFbAvgEonyfN6yWGLKc7zVjbpl2Y4CTjid9h2RfgPP6SEt6uHwEOply00yw==
  /@types/yargs/15.0.5:
    dependencies:
      '@types/yargs-parser': 15.0.0
    dev: false
    resolution:
      integrity: sha512-Dk/IDOPtOgubt/IaevIUbTgV7doaKkoorvOyYM2CMwuDyP89bekI7H4xLIwunNYiK9jhCkmc6pUrJk3cj2AB9w==
  /ansi-regex/5.0.0:
    dev: false
    engines:
      node: '>=8'
    resolution:
      integrity: sha512-bY6fj56OUQ0hU1KjFNDQuJFezqKdrAyFdIevADiqrWHwSlbmBNMHp5ak2f40Pm8JTFyM2mqxkG6ngkHO11f/lg==
  /ansi-styles/4.2.1:
    dependencies:
      '@types/color-name': 1.1.1
      color-convert: 2.0.1
    dev: false
    engines:
      node: '>=8'
    resolution:
      integrity: sha512-9VGjrMsG1vePxcSweQsN20KY/c4zN0h9fLjqAbwbPfahM3t+NL+M9HC8xeXG2I8pX5NoamTGNuomEUFI7fcUjA==
  /cliui/7.0.1:
    dependencies:
      string-width: 4.2.0
      strip-ansi: 6.0.0
      wrap-ansi: 7.0.0
    dev: false
    resolution:
      integrity: sha512-rcvHOWyGyid6I1WjT/3NatKj2kDt9OdSHSXpyLXaMWFbKpGACNW8pRhhdPUq9MWUOdwn8Rz9AVETjF4105rZZQ==
  /color-convert/2.0.1:
    dependencies:
      color-name: 1.1.4
    dev: false
    engines:
      node: '>=7.0.0'
    resolution:
      integrity: sha512-RRECPsj7iu/xb5oKYcsFHSppFNnsj/52OVTRKb4zP5onXwVF3zVmmToNcOfGC+CRDpfK/U584fMg38ZHCaElKQ==
  /color-name/1.1.4:
    dev: false
    resolution:
      integrity: sha512-dOy+3AuW3a2wNbZHIuMZpTcgjGuLU/uBL/ubcZF9OXbDo8ff4O8yVp5Bf0efS8uEoYo5q4Fx7dY9OgQGXgAsQA==
  /emoji-regex/8.0.0:
    dev: false
    resolution:
      integrity: sha512-MSjYzcWNOA0ewAHpz0MxpYFvwg6yjy1NG3xteoqz644VCo/RPgnr1/GGt+ic3iJTzQ8Eu3TdM14SawnVUmGE6A==
  /escalade/3.0.2:
    dev: false
    engines:
      node: '>=6'
    resolution:
      integrity: sha512-gPYAU37hYCUhW5euPeR+Y74F7BL+IBsV93j5cvGriSaD1aG6MGsqsV1yamRdrWrb2j3aiZvb0X+UBOWpx3JWtQ==
  /get-caller-file/2.0.5:
    dev: false
    engines:
      node: 6.* || 8.* || >= 10.*
    resolution:
      integrity: sha512-DyFP3BM/3YHTQOCUL/w0OZHR0lpKeGrxotcHWcqNEdnltqFwXVfhEBQ94eIo34AfQpo0rGki4cyIiftY06h2Fg==
  /is-fullwidth-code-point/3.0.0:
    dev: false
    engines:
      node: '>=8'
    resolution:
      integrity: sha512-zymm5+u+sCsSWyD9qNaejV3DFvhCKclKdizYaJUuHA83RLjb7nSuGnddCHGv0hk+KY7BMAlsWeK4Ueg6EV6XQg==
  /require-directory/2.1.1:
    dev: false
    engines:
      node: '>=0.10.0'
    resolution:
      integrity: sha1-jGStX9MNqxyXbiNE/+f3kqam30I=
  /string-width/4.2.0:
    dependencies:
      emoji-regex: 8.0.0
      is-fullwidth-code-point: 3.0.0
      strip-ansi: 6.0.0
    dev: false
    engines:
      node: '>=8'
    resolution:
      integrity: sha512-zUz5JD+tgqtuDjMhwIg5uFVV3dtqZ9yQJlZVfq4I01/K5Paj5UHj7VyrQOJvzawSVlKpObApbfD0Ed6yJc+1eg==
  /strip-ansi/6.0.0:
    dependencies:
      ansi-regex: 5.0.0
    dev: false
    engines:
      node: '>=8'
    resolution:
      integrity: sha512-AuvKTrTfQNYNIctbR1K/YGTR1756GycPsg7b9bdV9Duqur4gv6aKqHXah67Z8ImS7WEz5QVcOtlfW2rZEugt6w==
  /tslib/2.0.1:
    dev: false
    resolution:
      integrity: sha512-SgIkNheinmEBgx1IUNirK0TUD4X9yjjBRTqqjggWCU3pUEqIk3/Uwl3yRixYKT6WjQuGiwDv4NomL3wqRCj+CQ==
  /wrap-ansi/7.0.0:
    dependencies:
      ansi-styles: 4.2.1
      string-width: 4.2.0
      strip-ansi: 6.0.0
    dev: false
    engines:
      node: '>=10'
    resolution:
      integrity: sha512-YVGIj2kamLSTxw6NsZjoBxfSwsn0ycdesmc4p+Q21c5zPuZ1pl+NfxVdxPtdHvmNVOQ6XSYG4AUtyt/Fi7D16Q==
  /y18n/5.0.1:
    dev: false
    engines:
      node: '>=10'
    resolution:
      integrity: sha512-/jJ831jEs4vGDbYPQp4yGKDYPSCCEQ45uZWJHE1AoYBzqdZi8+LDWas0z4HrmJXmKdpFsTiowSHXdxyFhpmdMg==
  /yargs-parser/19.0.4:
    dev: false
    engines:
      node: '>=10'
    resolution:
      integrity: sha512-eXeQm7yXRjPFFyf1voPkZgXQZJjYfjgQUmGPbD2TLtZeIYzvacgWX7sQ5a1HsRgVP+pfKAkRZDNtTGev4h9vhw==
  /yargs/16.0.0:
    dependencies:
      cliui: 7.0.1
      escalade: 3.0.2
      get-caller-file: 2.0.5
      require-directory: 2.1.1
      string-width: 4.2.0
      y18n: 5.0.1
      yargs-parser: 19.0.4
    dev: false
    engines:
      node: '>=10'
    resolution:
      integrity: sha512-Ykb00VnWjee855QmeCrDAAmhVagt0T8PMML9WS2YrcU0VtvbeGq02MD7UiWmK6biiVPas6CaXmJNetL4Ye4+ng==
specifiers:
  '@tsfun/result': ^0.0.20
  '@types/node': ^14.6.4
  '@types/yargs': ^15.0.5
  tslib: ^2.0.1
  yargs: ^16.0.0
