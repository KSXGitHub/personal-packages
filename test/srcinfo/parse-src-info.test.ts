import path from 'path'
import { readFile } from 'fs-extra'
import { parseSrcInfo } from '@khai96x/srcinfo'

function loadAsset(basename: string) {
  const filename = path.resolve(__dirname, 'assets', basename)
  return readFile(filename, 'utf8')
}

it('valid', async () => {
  expect(parseSrcInfo(await loadAsset('valid.SRCINFO'))).toMatchSnapshot()
})

it('invalid', async () => {
  expect(parseSrcInfo(await loadAsset('invalid.SRCINFO'))).toMatchSnapshot()
})
