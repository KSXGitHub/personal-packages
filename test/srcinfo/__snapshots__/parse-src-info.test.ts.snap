// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid 1`] = `
Object {
  "error": Array [
    Object {
      "index": 2,
      "line": "	url : https://example.com",
      "message": "Line 3 is invalid",
      "type": "InvalidSyntax",
    },
    Object {
      "index": 3,
      "line": "	arch : x86_64",
      "message": "Line 4 is invalid",
      "type": "InvalidSyntax",
    },
    Object {
      "field": "pkgname",
      "message": "Field \\"pkgname\\" is required but not found",
      "type": "MissingRequiredField",
    },
    Object {
      "field": "pkgver",
      "message": "Field \\"pkgver\\" is required but not found",
      "type": "MissingRequiredField",
    },
    Object {
      "field": "pkgrel",
      "message": "Field \\"pkgrel\\" is required but not found",
      "type": "MissingRequiredField",
    },
  ],
  "tag": false,
}
`;

exports[`valid 1`] = `
Object {
  "tag": true,
  "value": Object {
    "arch": Array [
      "x86_64",
      "i686",
    ],
    "backup": Array [],
    "changelog": undefined,
    "checkdepends": Array [],
    "conflicts": Array [
      "foo",
      "foo-git",
      "bar",
      "bar-git",
    ],
    "depends": Array [],
    "epoch": undefined,
    "groups": Array [],
    "install": undefined,
    "license": Array [
      "MIT",
      "MPL",
    ],
    "makedepends": Array [
      "coreutils",
      "ar",
      "tar",
    ],
    "md5sums": Array [],
    "noextract": Array [],
    "optdepends": Array [],
    "options": Array [],
    "pkgdesc": "An example .SRCINFO",
    "pkgname": "foo-bin",
    "pkgrel": "3",
    "pkgver": "0.1.2",
    "provides": Array [
      "foo",
      "bar",
    ],
    "replaces": Array [],
    "sha1sums": Array [],
    "sha224sums": Array [],
    "sha256sums": Array [],
    "sha348sums": Array [],
    "sha512sums": Array [
      "SKIP",
      "SKIP",
      "SKIP",
    ],
    "source": Array [
      "https://example.com/a.tar.gz",
      "https://example.com/b.tar.gz",
      "https://example.com/c.tar.gz",
    ],
    "url": "https://example.com",
    "validgpgkeys": Array [],
  },
}
`;
